Fix a bug in NameSimplifier.

hrkrshnn opened the pull request:
A new name created by NameSimplifier could also be created by NameDispenser, since the knowledge of
the new name was not stored in NameSimplifier.

hrkrshnn committed:
Fix a bug in NameSimplifier.

A new name created by NameSimplifier could also be created by NameDispenser, since the knowledge of
the new name was not stored in NameSimplifier.

cameel reviewed (commented):
Looks good.

Can you add a test that would fail before this change?

hrkrshnn commented:
> Can you add a test that would fail before this change?

@cameel That won't be easy. Let me try to see if I can create one. I think the failures in https://github.com/ethereum/solidity/pull/10418 are because of this. The tests there would pass after making this change.

chriseth commented on libyul/optimiser/NameSimplifier.h:
```diff
@@ -47,7 +48,10 @@ class NameSimplifier: public ASTModifier
+		NameSimplifier nameSimplifier{_context, _ast};
+		nameSimplifier(_ast);
+		for (auto const& names: nameSimplifier.m_translations)
+			nameSimplifier.m_context.dispenser.markUsed(names.second);
```
I think this should be part of `NameSimplifier::findSimplification` instead (and that function could be split up a little).

chriseth commented on libyul/optimiser/NameSimplifier.cpp:
```diff
@@ -109,8 +110,10 @@ void NameSimplifier::findSimplification(YulString _name)
-		m_usedNames.insert(YulString(name));
-		m_translations[_name] = YulString(name);
+		YulString newName{name};
+		m_usedNames.insert(newName);
```
Do we actually still need m_usedNames if the dispenser keeps track?

Also, in general: Does the dispenser mark names as available again if they are deleted?

hrkrshnn replied:
> Do we actually still need m_usedNames if the dispenser keeps track?

Let me check.

> Also, in general: Does the dispenser mark names as available again if they are deleted?

No we don't do that. I think it's better if we don't do this.

chriseth replied:
At least at the end we should do it. If we change the name simplifier to run once at the end (in addition to once at the beginning) it should be able to re-use names that were used once during the optimization process but are not used at the end.

chriseth commented:
Actually I think we should take the name simplifier out of the steps available to users.

hrkrshnn committed:
NameDispenser: Added three public functions.

The private function `illegalName` was converted into a public function.

A new public getter function `usedNames()` returns a reference to `m_usedNames`.

A public function reset that would change all `m_usedNames` to *only* the names that are used in the
AST. Also resets the counter. This will be used before the final NameSimplifier step in the
optimization phase.

The first two functions were added so that `NameSimplifier` could use it instead of traversing the AST
using `NameCollector` to find used names.

hrkrshnn committed:
NameSimplifier is removed from steps available to users.

hrkrshnn commented:
I did some refactoring that removes `m_usedNames` and make the step rely on `NameDispenser`. Also removed NameSimplifier from steps available to users.

chriseth commented on libyul/optimiser/Suite.cpp:
```diff
@@ -140,6 +141,8 @@ void OptimiserSuite::run(
 		if (ast.statements.size() > 1 && std::get<Block>(ast.statements.front()).statements.empty())
 			ast.statements.erase(ast.statements.begin());
 	}
+
```
Don't you think it makes sense to "reset" the name dispenser at this point?

hrkrshnn replied:
I added a reset.

chriseth commented on libyul/optimiser/NameDispenser.cpp:
```diff
@@ -61,3 +62,9 @@ bool NameDispenser::illegalName(YulString _name)
 	return isRestrictedIdentifier(m_dialect, _name) || m_usedNames.count(_name);
 }
+
+void NameDispenser::reset(Block const& _ast)
```
I think instead of adding a `reset` function it might be more idiomatic to re-assign `suite.m_dispenser` from a fresh object instead.

hrkrshnn replied:
I did think about that, but `NameDispenser` is not really copyable/assignable because of the member `Dialect const& m_dialect`. Not sure if changing it to a pointer now is worth the hassle. What do you think?

chriseth replied:
We could make OptimiserSuite::m_dispenser a unique_ptr but keep OptiimserStepContext::dispenser a reference. This should require few changes, does it?

hrkrshnn replied:
> We could make OptimiserSuite::m_dispenser a unique_ptr but keep OptiimserStepContext::dispenser a reference. This should require few changes, does it?

No, we have to make both of them into a pointer. Because the problem here is that the steps need to use the reference `OptimiserStepContext::dispenser`. So if we assign a new Dispenser to the `unique_ptr OptimizerSuite::m_dispenser`, then the reference should also be changed. So, we'll have to make both of them into a pointer.  I'd prefer a reset function over making this change.

chriseth replied:
Oh right, of course!

chriseth commented on libyul/optimiser/NameSimplifier.cpp:
```diff
@@ -28,19 +30,13 @@
-	set<YulString> allNames = NameCollector(_ast).names();
-	m_usedNames += allNames;
-	for (YulString name: allNames)
+	for (YulString const& name: m_context.dispenser.usedNames())
```
This assumes that the dispenser has been reset right before - I'm not sure we should make this assumption.

hrkrshnn replied:
This just assumes that Dispenser is initialized, there would be no issues if it was reset before or not.

Maybe it's just better to use `NameCollector(_ast).names()` it would make more sense. I'll fix it.

hrkrshnn replied:
Fixed this.

hrkrshnn commented on Changelog.md:
```diff
@@ -12,6 +12,8 @@ Compiler Features:
 Bugfixes:
  * SMTChecker: Fix internal compiler error when doing bitwise compound assignment with string literals.
+ * Yul Optimizer: Fix a bug in NameSimplifier where a new name created by NameSimplifier could also be created by NameDispenser.
+ * Yul Optimizer: Removed NameSimplifier from optimization steps available to users.
```
Added an extra line in changelog.
